lib/fs/filesystem.go

```diff
@@ -246,18 +236,23 @@ func NewFilesystem(fsType FilesystemType, uri string, opts ...Option) Filesystem
 	}
 	opts = opts[:i]
 
+	// Construct file system using the registered factory function
 	var fs Filesystem
-	switch fsType {
-	case FilesystemTypeBasic:
-		fs = newBasicFilesystem(uri, opts...)
-	case FilesystemTypeFake:
-		fs = newFakeFilesystem(uri, opts...)
-	default:
-		l.Debugln("Unknown filesystem", fsType, uri)
+	if fsFactory, ok := filesystemFactories[fsType]; ok {
```

Comment:
I think this thing could be slightly simplified by something like

```
var fs Filesystem
var err error
if fact, ok := filesystemFactories[fsType]; ok {
  fs, err = fsFactory(fsType, uri, opts...)
} else {
  err = fmt.Errorf("File system type '%s' not recognized", fsType)
}
if err != nil {
  fs = &errorFilesystem{
    fsType: fsType,
    uri:    uri,
    err:    err,
  }
}
```

oh and we must also lock the mutex here